backoff.go

```diff
@@ -0,0 +1,71 @@
+package pubsub
+
+import (
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const (
+	MinBackoffDelay   = 100 * time.Millisecond
+	MaxBackoffDelay   = 10 * time.Second
+	TimeToLive        = 10 * time.Minute
+	BackoffMultiplier = 2
+)
+
+type backoffHistory struct {
+	duration  time.Duration
+	lastTried time.Time
+}
+
+type backoff struct {
+	mu   sync.Mutex
+	info map[peer.ID]*backoffHistory
+	ct   int // size threshold that kicks off the cleaner
+}
+
+func newBackoff(sizeThreshold int) *backoff {
+	return &backoff{
+		mu:   sync.Mutex{},
+		ct:   sizeThreshold,
+		info: make(map[peer.ID]*backoffHistory),
+	}
+}
+
+func (b *backoff) updateAndGet(id peer.ID) time.Duration {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	h, ok := b.info[id]
+	if !ok || time.Since(h.lastTried) > TimeToLive {
+		// first request goes immediately.
+		h = &backoffHistory{
+			duration: time.Duration(0),
+		}
+	} else if h.duration < MinBackoffDelay {
+		h.duration = MinBackoffDelay
+	} else if h.duration < MaxBackoffDelay {
+		h.duration = time.Duration(BackoffMultiplier * h.duration)
```

Comment:
can we add some jitter?

Reply:
https://github.com/libp2p/go-libp2p-pubsub/pull/483/commits/a77d4355e47d85fc0e7cb6373f9e1927e7add3de